src/protocol/ncc1701_nrf24l01.c

```diff
@@ -0,0 +1,402 @@
+    NRF24L01_WriteReg(NRF24L01_01_EN_AA, 0x00);             // No Auto Acknowledgment on all data pipes
+    NRF24L01_WriteReg(NRF24L01_02_EN_RXADDR, 0x01);         // Enable data pipe 0 only
+    NRF24L01_WriteReg(NRF24L01_11_RX_PW_P0, NCC_RX_PACKET_LEN); // Enable rx pipe 0
+    NRF24L01_SetBitrate(NRF24L01_BR_250K);                  // NRF24L01_BR_1M, NRF24L01_BR_2M, NRF24L01_BR_250K
+    NRF24L01_SetPower(tx_power);
+    NRF24L01_FlushRx();
+    NRF24L01_WriteReg(NRF24L01_00_CONFIG, (0 << NRF24L01_00_EN_CRC)   // switch to TX mode and disable CRC
+                                        | (1 << NRF24L01_00_CRCO)
+                                        | (1 << NRF24L01_00_PWR_UP)
+                                        | (0 << NRF24L01_00_PRIM_RX));
+}
+
+static const u8 NCC_xor[]={0x80, 0x44, 0x64, 0x75, 0x6C, 0x71, 0x2A, 0x36, 0x7C, 0xF1, 0x6E, 0x52, 0x09, 0x9D};
+static void NCC_Crypt_Packet()
+{
+    u16 crc = 0;
+    for(u8 i=0; i < NCC_TX_PACKET_LEN-2; i++)
+    {
+        packet[i] ^= NCC_xor[i];
+        crc = crc16_update(crc, packet[i], 8);
+    }
+    crc ^= 0x60DE;
+    packet[NCC_TX_PACKET_LEN-2] = crc >> 8;
+    packet[NCC_TX_PACKET_LEN-1] = crc;
+}
+
+static u8 NCC_Decrypt_Packet()
+{
+    u16 crc = 0;
+    for(u8 i=0; i < NCC_RX_PACKET_LEN-2; i++)
+    {
+        crc = crc16_update(crc, packet[i], 8);
+        packet[i] ^= NCC_xor[i];
+    }
+    crc ^= 0xA950;
+    if((crc >> 8) == packet[NCC_RX_PACKET_LEN-2] && (crc & 0xFF) == packet[NCC_RX_PACKET_LEN-1] )
+    {// CRC match
+        return 1;
+    }
+    return 0;
+}
+
+#define CHAN_RANGE (CHAN_MAX_VALUE - CHAN_MIN_VALUE)
+static u8 scale_channel(u8 ch, u8 destMin, u8 destMax)
+{
+    s32 chanval = Channels[ch];
+    s32 range = (s32) destMax - (s32) destMin;
+
+    if (chanval < CHAN_MIN_VALUE)
+        chanval = CHAN_MIN_VALUE;
+    else if (chanval > CHAN_MAX_VALUE)
+        chanval = CHAN_MAX_VALUE;
+    return (range * (chanval - CHAN_MIN_VALUE)) / CHAN_RANGE + destMin;
+}
+
+#define GET_FLAG(ch, mask) (Channels[ch] > 0 ? mask : 0)
+static void NCC_Write_Packet()
+{
+    packet[0] = 0xAA;
+    packet[1] = rx_tx_addr[0];
+    packet[2] = rx_tx_addr[1];
+    packet[3] = rx_id[0];
+    packet[4] = rx_id[1];
+    packet[5] = scale_channel(CHANNEL3, 0, 0x3d); // throttle 00-3D
+    packet[6] = scale_channel(CHANNEL2, 0, 0xff); // elevator, original: 61-80-9F but works with 00-80-FF
+    packet[7] = scale_channel(CHANNEL1, 0xff, 0); // aileron, original: 61-80-9F but works with 00-80-FF
+    packet[8] = scale_channel(CHANNEL4, 0xff, 0); // rudder, original: 61-80-9F but works with 00-80-FF
+    packet[9] = rx_id[2];
+    packet[10] = rx_id[3];
+    packet[11] = rx_id[4];
+    packet[12] = GET_FLAG(CHANNEL_WARP, 0x02);    // Warp:0x00 -> 0x02
+    packet[13] = packet[5]+packet[6]+packet[7]+packet[8]+packet[12];
+    if(phase == NCC_BIND_TX1)
+    {
+        packet[0] = 0xBB;
+        packet[5] = 0x01;
+        packet[6] = rx_tx_addr[2];
```

Comment:
can u reformat the code a little bit better? looks like when bind, you only need first 7 bytes. other protocol has the functions like build_bind_packet and build_data_packet.

Reply:
it's ported from the MULTI project and I prefer keeping the code as it is, it will be easier to maintain.